server/services/store/sqlstore/blocks.go

```diff
@@ -788,3 +802,264 @@ func (s *SQLStore) duplicateBlock(db sq.BaseRunner, boardID string, blockID stri
+	if boardID == "" {
+		return ErrEmptyBoardID{}
+	}
+
+	where := fmt.Sprintf("board_id='%s'", boardID)
+	if parentID != "" {
+		where += fmt.Sprintf(" AND parent_id='%s'", parentID)
+	}
+
+	/*
+		SELECT * FROM focalboard_blocks_history AS bh,
+			(SELECT id, max(insert_at) AS max_insert_at FROM focalboard_blocks_history WHERE board_id='xxx' AND parent_id='yyy' GROUP BY id) AS sub
+		WHERE bh.id = sub.id AND bh.insert_at=sub.max_insert_at AND bh.delete_at <> 0
+	*/
+	selectQuery := s.getQueryBuilder(db).
+		Select(
+			"bh.board_id",
+			"'' AS channel_id",
+			"bh.id",
+			"bh.parent_id",
+			"bh.schema",
+			"bh.type",
+			"bh.title",
+			"bh.fields",
+			"'"+modifiedBy+"' AS modified_by",
+			"bh.create_at",
+			s.castInt(utils.GetMillis(), "update_at"),
+			s.castInt(0, "delete_at"),
+			"bh.created_by",
+		).
+		From(fmt.Sprintf(`
+				%sblocks_history AS bh,
+				(SELECT id, max(insert_at) AS max_insert_at FROM %sblocks_history WHERE %s GROUP BY id) AS sub`,
+			s.tablePrefix, s.tablePrefix, where)).
+		Where("bh.id=sub.id").
+		Where("bh.insert_at=sub.max_insert_at").
+		Where(sq.NotEq{"bh.delete_at": 0})
+
+	columns := []string{
+		"board_id",
+		"channel_id",
+		"id",
+		"parent_id",
+		s.escapeField("schema"),
+		"type",
+		"title",
+		"fields",
+		"modified_by",
+		"create_at",
+		"update_at",
+		"delete_at",
+		"created_by",
+	}
+
+	insertQuery := s.getQueryBuilder(db).Insert(s.tablePrefix + "blocks").
+		Columns(columns...).
+		Select(selectQuery)
+
+	insertHistoryQuery := s.getQueryBuilder(db).Insert(s.tablePrefix + "blocks_history").
+		Columns(columns...).
+		Select(selectQuery)
+
+	sql, args, err := insertQuery.ToSql()
+	s.logger.Trace("undeleteBlockChildren - insertQuery",
+		mlog.String("sql", sql),
+		mlog.Array("args", args),
+		mlog.Err(err),
+	)
+
+	sql, args, err = insertHistoryQuery.ToSql()
+	s.logger.Trace("undeleteBlockChildren - insertHistoryQuery",
+		mlog.String("sql", sql),
+		mlog.Array("args", args),
+		mlog.Err(err),
+	)
+
+	// insert into blocks table must happen before history table, otherwise the history
+	// table will be changed and the second query will fail to find the same records.
+	result, err := insertQuery.Exec()
+	if err != nil {
+		return err
+	}
+	rowsAffected, _ := result.RowsAffected()
+	s.logger.Debug("undeleteBlockChildren - insertQuery", mlog.Int64("rows_affected", rowsAffected))
+
+	result, err = insertHistoryQuery.Exec()
+	if err != nil {
+		return err
+	}
+	rowsAffected, _ = result.RowsAffected()
+	s.logger.Debug("undeleteBlockChildren - insertHistoryQuery", mlog.Int64("rows_affected", rowsAffected))
+
+	return nil
+}
+
+func (s *SQLStore) findOrphansForBoards(db sq.BaseRunner, limit int) ([]string, error) {
+	if limit <= 0 {
+		limit = 1000
+	}
+	if limit > 1000 {
+		return nil, ErrLimitExceeded{1000}
+	}
+
+	// Find all orphaned child blocks for deleted boards.
+	// Squirrel can't express multiple tables in FROM, nor can it mix `From` and `FromSelect`.
+	/*
```

Comment:
The query shown here, I think is misleading, the comment on top, on the other hand, I think is right.